// @tsparticles/updater-opacity/umd/OpacityUpdater.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@tsparticles/engine", "./Utils.js"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.OpacityUpdater = void 0;
    const engine_1 = require("@tsparticles/engine");
    const Utils_js_1 = require("./Utils.js");
    class OpacityUpdater {
        constructor(container) {
            this.container = container;
        }
        init(particle) {
            const opacityOptions = particle.options.opacity;
            particle.opacity = (0, engine_1.initParticleNumericAnimationValue)(opacityOptions, 1);
            const opacityAnimation = opacityOptions.animation;
            if (opacityAnimation.enable) {
                particle.opacity.velocity =
                    ((0, engine_1.getRangeValue)(opacityAnimation.speed) / 100) * this.container.retina.reduceFactor;
                if (!opacityAnimation.sync) {
                    particle.opacity.velocity *= (0, engine_1.getRandom)();
                }
            }
        }
        isEnabled(particle) {
            return (!particle.destroyed &&
                !particle.spawning &&
                !!particle.opacity &&
                particle.opacity.enable &&
                ((particle.opacity.maxLoops ?? 0) <= 0 ||
                    ((particle.opacity.maxLoops ?? 0) > 0 &&
                        (particle.opacity.loops ?? 0) < (particle.opacity.maxLoops ?? 0))));
        }
        reset(particle) {
            if (particle.opacity) {
                particle.opacity.time = 0;
                particle.opacity.loops = 0;
            }
        }
        update(particle, delta) {
            if (!this.isEnabled(particle)) {
                return;
            }
            (0, Utils_js_1.updateOpacity)(particle, delta);
        }
    }
    exports.OpacityUpdater = OpacityUpdater;
});
