// @tsparticles/plugin-emitters-shape-circle/cjs/EmittersCircleShape.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmittersCircleShape = void 0;
const plugin_emitters_1 = require("@tsparticles/plugin-emitters");
const engine_1 = require("@tsparticles/engine");
class EmittersCircleShape extends plugin_emitters_1.EmitterShapeBase {
    constructor(position, size, fill, options) {
        super(position, size, fill, options);
    }
    async init() {
    }
    async randomPosition() {
        const size = this.size, fill = this.fill, position = this.position, generateTheta = (x, y) => {
            const u = (0, engine_1.getRandom)() / 4.0, theta = Math.atan((y / x) * Math.tan(2 * Math.PI * u)), v = (0, engine_1.getRandom)();
            if (v < 0.25) {
                return theta;
            }
            else if (v < 0.5) {
                return Math.PI - theta;
            }
            else if (v < 0.75) {
                return Math.PI + theta;
            }
            else {
                return -theta;
            }
        }, radius = (x, y, theta) => (x * y) / Math.sqrt((y * Math.cos(theta)) ** 2 + (x * Math.sin(theta)) ** 2), [a, b] = [size.width / 2, size.height / 2], randomTheta = generateTheta(a, b), maxRadius = radius(a, b, randomTheta), randomRadius = fill ? maxRadius * Math.sqrt((0, engine_1.getRandom)()) : maxRadius;
        return {
            position: {
                x: position.x + randomRadius * Math.cos(randomTheta),
                y: position.y + randomRadius * Math.sin(randomTheta),
            },
        };
    }
}
exports.EmittersCircleShape = EmittersCircleShape;
