// @tsparticles/shape-text/browser/TextDrawer.js
import { executeOnSingleOrMultiple, isInArray, itemFromSingleOrMultiple, loadFont, } from "@tsparticles/engine";
export const validTypes = ["text", "character", "char", "multiline-text"];
export class TextDrawer {
    constructor() {
        this._drawLine = (context, line, radius, opacity, index, fill) => {
            const offsetX = (line.length * radius) / 2, pos = {
                x: -offsetX,
                y: radius / 2,
            }, diameter = radius * 2;
            if (fill) {
                context.fillText(line, pos.x, pos.y + diameter * index);
            }
            else {
                context.strokeText(line, pos.x, pos.y + diameter * index);
            }
        };
    }
    draw(data) {
        const { context, particle, radius, opacity } = data, character = particle.shapeData;
        if (!character) {
            return;
        }
        const textData = character.value;
        if (textData === undefined) {
            return;
        }
        if (particle.text === undefined) {
            particle.text = itemFromSingleOrMultiple(textData, particle.randomIndexData);
        }
        const text = particle.text, style = character.style ?? "", weight = character.weight ?? "400", size = Math.round(radius) * 2, font = character.font ?? "Verdana", fill = particle.shapeFill;
        const lines = text?.split("\n");
        if (!lines) {
            return;
        }
        context.font = `${style} ${weight} ${size}px "${font}"`;
        context.globalAlpha = opacity;
        for (let i = 0; i < lines.length; i++) {
            this._drawLine(context, lines[i], radius, opacity, i, fill);
        }
        context.globalAlpha = 1;
    }
    async init(container) {
        const options = container.actualOptions;
        if (validTypes.find((t) => isInArray(t, options.particles.shape.type))) {
            const shapeOptions = validTypes
                .map((t) => options.particles.shape.options[t])
                .find((t) => !!t), promises = [];
            executeOnSingleOrMultiple(shapeOptions, (shape) => {
                promises.push(loadFont(shape.font, shape.weight));
            });
            await Promise.all(promises);
        }
    }
    particleInit(container, particle) {
        if (!particle.shape || !validTypes.includes(particle.shape)) {
            return;
        }
        const character = particle.shapeData;
        if (character === undefined) {
            return;
        }
        const textData = character.value;
        if (textData === undefined) {
            return;
        }
        particle.text = itemFromSingleOrMultiple(textData, particle.randomIndexData);
    }
}
