// @tsparticles/plugin-emitters-shape-square/browser/EmittersSquareShape.js
import { EmitterShapeBase } from "@tsparticles/plugin-emitters";
import { getRandom } from "@tsparticles/engine";
function randomSquareCoordinate(position, offset) {
    return position + offset * (getRandom() - 0.5);
}
export class EmittersSquareShape extends EmitterShapeBase {
    constructor(position, size, fill, options) {
        super(position, size, fill, options);
    }
    async init() {
    }
    async randomPosition() {
        const fill = this.fill, position = this.position, size = this.size;
        if (fill) {
            return {
                position: {
                    x: randomSquareCoordinate(position.x, size.width),
                    y: randomSquareCoordinate(position.y, size.height),
                },
            };
        }
        else {
            const halfW = size.width / 2, halfH = size.height / 2, side = Math.floor(getRandom() * 4), v = (getRandom() - 0.5) * 2;
            switch (side) {
                case 0:
                    return {
                        position: {
                            x: position.x + v * halfW,
                            y: position.y - halfH,
                        },
                    };
                case 1:
                    return {
                        position: {
                            x: position.x - halfW,
                            y: position.y + v * halfH,
                        },
                    };
                case 2:
                    return {
                        position: {
                            x: position.x + v * halfW,
                            y: position.y + halfH,
                        },
                    };
                case 3:
                default:
                    return {
                        position: {
                            x: position.x + halfW,
                            y: position.y + v * halfH,
                        },
                    };
            }
        }
    }
}
